src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -1525,3 +1525,74 @@ case class MatrixRowsTable(child: MatrixIR) extends TableIR {
       mv.rowsRVD())
   }
 }
+
+case class MatrixFilterEntries(child: MatrixIR, pred: IR) extends MatrixIR {
+  val children: IndexedSeq[BaseIR] = Array(child, pred)
+
+  def copy(newChildren: IndexedSeq[BaseIR]): MatrixFilterEntries = {
+    assert(newChildren.length == 2)
+    MatrixFilterEntries(newChildren(0).asInstanceOf[MatrixIR], newChildren(1).asInstanceOf[IR])
+  }
+
+  val typ: MatrixType = child.typ
+
+  def execute(hc: HailContext): MatrixValue = {
+    val mv = child.execute(hc)
+
+    val localGlobalType = child.typ.globalType
+    val globalsBc = mv.globals.broadcast
+    val localColValuesType = TArray(child.typ.colType)
+    val colValuesBc = mv.colValues.broadcast
+
+    val colValuesType = TArray(child.typ.colType)
+
+    val entry = "entry"
+    val x = ir.InsertFields(ir.Ref("va", child.typ.rvRowType),
```

Comment:
Yeah, I agree, but it isn't easily fixed right now.  The names are determined by the EvalContexts which are all over the place, need to work for both AST and IR for the moment, and are fixed to use global, va, sa, g.  Second, you are totally correct that we're playing a trick by treating the entire row as "va" but the alternative is quite painful (re-construct the row and pull out the gs as separate fields, do the transformation, and reassemble).  So I think we need to leave this for now but I am happy to discuss alternative solutions.

In the long term, we should "lower" MatrixTable operations (which never refer to the row explicitly, because that's an implementation detail) to Table operations (where the row has the entry field in the usual way) and everything there is explicit.  I think this fixes the problem in the long term.